### ISSUE TITLE
standardize on behavior of 0 when specifying geometries

### ISSUE THREAD

**Issue by dankamongmen** [dankamongmen/notcurses#1696]
A number of functions take geometries, with `ncplane_as_rgba()` as an example. Some take an upper-left and lower-right point; some take an origin and lengths. This disparity is bad enough, but then we also sometimes support 0,0 to indicate "everything" and sometimes require -1,-1 for same, with 0,0 being an error.

I'd like to normalize these before ABI3. Let's get a list of all entry points taking a geometry as input, and see if there are any which require/prefer one or the other form. If not (I can't imagine why they would), let's figure out whether 0,0 ought be an error, or "everything left", and the same for negative values.

Regarding this last, there's a discontinuity when we go to "everything" regardless of whether it's at -1,-1 or 0,0, since 1,1 gets at most 1 cell, and it's been decreasing from the maximum to get there. If all the negatives work, and 0,0 does as well (or if both classes are errors), there are only two total discontinuities. If one or the other works, there are 3. If the caller is using unsigneds and coercing them to ints, it makes more sense to consider all negatives an error, imho. If the caller is writing a natural algorithm, one that terminates at 0,0, it makes more sense to consider 0,0 an error imho. But such a caller would then need to check for the error, in which case they could just check for 0,0, and avoid the call. From a convenience perspective, having 0,0 be "everything left" seems more attrative imho than -1,-1, but the biggest issue for convenience is probably simply consistent handling of arguments.

I'm inclined to think negatives ought always be errors, so that overflow due to truncated unsigned parameters is always caught. If that's what we do, convenience demands that the 0,0 option remain for "everything else". So I'm inclined to handle it that way:

 - one or two negative arguments is an error
 - one zero argument is an error
 - two zero arguments means "specified origin through lower right"
 - two positive arguments means "specified area"
 - origin outside the target is an error
 - lower right, when specified, outside the target is an error

**Comment by dankamongmen** [comment]
relevant functions: `ncplane_resize()`, `ncplane_resize_simple()`, `ncplane_contents()`, `ncplane_hline_interp()`, `ncplane_vline_interp()`, `ncplane_box()`, `ncplane_gradient()`, `ncplane_highgradient()`, `ncplane_gradient_sized()`, `ncplane_highgradient_sized()`, `ncplane_mergedown()`, `ncplane_erase_region()`, `ncvisual_from_plane()`, `ncvisual_render()`, `ncplane_as_rgba()`, and their direct mode relatives.

**Comment by dankamongmen** [comment]
Here's another question: `keepy` and `keepx` (wrt `ncplane_resize()`) must be within the old plane, but `keeplenx` and `keepleny` only need be within the new one. why wouldn't `keepleny + keepy` and `keeplenx + keepx` need to be within the old plane, as well? this leads to output like:

 ```
ncplane_resize_internal:742:70x80 @ 0/0 → 70/80 @ 0/0 (keeping 70x80 from 0/0)
ncplane_resize_internal:742:14x80 @ 0/0 → 14/80 @ 0/0 (keeping 14x80 from 0/0)
```

which looks wrong and jarring (how can i keep 70x80 from 0x0?), even if it's not actually leading to a false copy.

**Comment by dankamongmen** [comment]
realizing that `visual_blit` is also involved here locks in 0 as full length, given that we don't want to invalidate every default-initialized `ncvisual_options` in existence.

**Comment by dankamongmen** [comment]
so we need change:
* resize: move to unsigned
* contents: 0 for all, not -1
* hline: 0 ought be error, move to unsigned
* vline: 0 ought be error, move to unsigned
* box: move to unsigned, 0 size is error, 0 for all
* box_sized: move to unsigned, 0 for all
* gradient: same as box
* mergedown: 0 for all
* from_plane: 0 for all
* as_rgba: 0 for all, unsigned

do not change:
* erase_region
* visual_blit

basically if you're area (lengths), you ought be using unsigned, and 0 ought indicate all in that dimension (always right and down). if you're using endpoints (coordinates), you ought be using signed, it ought work in both directions, and there ought not be any special values. let's get this done for 2.4.9.

**Comment by dankamongmen** [comment]
ugh i really hate all this churn for something of such questionable value. you've gotta figure at least one subtle bug snuck in on this branch. is it worth it?
